plugins/soundsourcem4a/soundsourcem4a.cpp

```diff
@@ -561,33 +553,50 @@ SINT SoundSourceM4A::readSampleFrames(
         const SINT numberOfSamplesDecoded = decFrameInfo.samples;
         DEBUG_ASSERT(numberOfSamplesDecoded <= decodeBufferCapacity);
         SINT numberOfSamplesRead;
-        if (pDecodeBuffer == pSampleBuffer) {
-            numberOfSamplesRead = math_min(numberOfSamplesDecoded, numberOfSamplesRemaining);
-            pSampleBuffer += numberOfSamplesRead;
+        if (writableSampleFrames.writableData() &&
+                (pDecodeBuffer == writableSampleFrames.writableData(outputSampleOffset))) {
+            // Decoded in-place
+            DEBUG_ASSERT(numberOfSamplesDecoded <= numberOfSamplesRemaining);
+            numberOfSamplesRead = numberOfSamplesDecoded;
+            outputSampleOffset += numberOfSamplesRead;
         } else {
-            m_sampleBuffer.readFromTail(decodeBufferCapacity - numberOfSamplesDecoded);
-            const SampleBuffer::ReadableChunk readableChunk(
-                    m_sampleBuffer.readFromHead(numberOfSamplesRemaining));
-            numberOfSamplesRead = readableChunk.size();
-            if (pSampleBuffer) {
-                SampleUtil::copy(pSampleBuffer, readableChunk.data(), numberOfSamplesRead);
-                pSampleBuffer += numberOfSamplesRead;
+            // Decoded into temporary buffer
+            DEBUG_ASSERT(numberOfSamplesDecoded <= decodeBufferCapacity);
+            // Shrink the size of the buffer to the samples that have
+            // actually been decoded, i.e. dropping unneeded samples
+            // from the back of the buffer.
+            m_sampleBuffer.dropFromTail(decodeBufferCapacity - numberOfSamplesDecoded);
```

Comment:
Lets name the function to what it does here. It adjusts the readable range to the just written samples  
m_sampleBuffer.shrinkToWritten(numberOfSamplesDecoded);
m_sampleBuffer.confirmSamplesWritten(numberOfSamplesDecoded);